qiskit_experiments/warnings.py

```diff
@@ -0,0 +1,245 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A collection of typical warnings."""
+
+import functools
+import warnings
+
+from typing import Callable, Optional, Type, Dict
+
+
+def deprecated_logic(
+    version_removed: Optional[str] = None,
+    use_instead: Optional[str] = None,
+) -> Callable:
+    """A function or method decorator to show deprecation warning.
+
+    Args:
+        version_removed: The Qiskit Experiment version that this function is removed.
+        use_instead: Alternative approach.
+
+    Examples:
+
+        .. code-block::
+
+            @deprecated_logic(version_removed="0.3", use_instead="use new_function")
+            def old_function(*args, **kwargs):
+                pass
+
+            def new_function(*args, **kwargs):
+                pass
+
+    Returns:
+        Deprecated function or method.
+    """
+
+    def deprecated_wrapper(func: Callable):
+        @functools.wraps(func)
+        def _wrap(*args, **kwargs):
+            namespace = func.__qualname__.split(".")
+            if len(namespace) == 1:
+                message = f"The function '{func.__name__}' has been deprecated and "
+            else:
+                cls_name, meth_name = namespace
+                message = f"The method '{meth_name}' of '{cls_name}' class has been deprecated and "
+            if version_removed:
+                message += f"will be removed in Qiskit Experiments {version_removed}. "
+            else:
+                message += "will be removed in future release. "
+            if use_instead:
+                message += f"Please '{use_instead}' instead. "
```

Comment:
Here this hosuld just be
```
if msg:
    message += msg
```
to allow more flexibility for setting custom messages